Add Aviate tutorial for leveraging Aviate catalog and metering APIs to recreate a AI billing use case. (#609)
killbill/killbill-docs: userguide/aviate/aviate-usage-tutorial.adoc

= Usage Tutorial - AI Use Case

include::{sourcedir}/aviate/includes/aviate-card.adoc[]

== Introduction

This tutorial shows how to use the Aviate APIs together with open-source Kill Bill APIs to address usage-based billing scenarios.

Kill Bill already supports sophisticated catalogs, usage-based plans, and automated invoicing. Aviate extends these capabilities with:

> * A scalable metering system for capturing usage data points.

Reviewer: We should update the link from https://github.com/killbill/killbill-docs/blob/f8dcb502fc963bdca102daaa2721924d5c7adf69/userguide/tutorials/consumable_in_arrear.adoc?plain=1#L124 to point to Aviate docs instead.

* Wallet functionality integrated with Kill Bill to manage credit pools and balances.
* Advanced product and plan management, also integrated with Kill Bill.

In this example, we simulate an AI use cases where input tokens are recorded as usage and billed periodically via paid credits.

== Prerequisites

You will need:

* A running Kill Bill with the Aviate plugin installed.

> * A valid API access tokens for all Aviate APIs calls - referred to is `ID_TOKEN`

Reviewer: ```suggestion
* A valid API access tokens for all Aviate APIs calls - referred to as `ID_TOKEN`
```

== Setup: Configuring a Tenant

=== Step 1: Create a tenant

We start by creating a tenant `usage-scenario`.

> Use the standard Kill Bill API: https://killbill.github.io/slate/tenant.html#create-a-tenant

Reviewer: ```suggestion
Use the standard Kill Bill API: https://apidocs.killbill.io/tenant.html#create-a-tenant
```

Author: I have not changed those because this is the link I get when I click from our doc - and editing all links manually afterwards seems error prone.

[source,bash]
-----
curl -v -X POST -u admin:password \
-H "Content-Type: application/json" \
-H "Accept: application/json" \
-H "X-Killbill-CreatedBy: demo" \
-d '{ "apiKey": "usage-scenario", "apiSecret": "usage-scenario" }' \
"http://127.0.0.1:8080/1.0/kb/tenants"
-----

=== Step 2: Create a billing meter

> A BillingMeter defines how usage points are filtered and aggregated. Before creating a Plan with usage, a valid billing meter must exist. This is a new abstraction in Aviate that extends the default capabilities from Kill Bill open-source.

Reviewer: ```suggestion
A BillingMeter defines how usage points are filtered and aggregated. Before creating a Plan with usage, a valid billing meter must exist. This is an Aviate specific abstraction that extends the default capabilities from Kill Bill open-source.
```

-H 'X-killbill-apisecret: usage-scenario' \
-d '[{"name": "Input Tokens (per Million)", "code": "input-M-tokens", "eventKey": "input.tokens", "aggregationType": "SUM"}]' \
http://127.0.0.1:8080/plugins/aviate-plugin/v1/metering/billingMeters
-----

=== Step 3: Create a simple catalog with one Plan

The Plan includes a usage section linked to the billing meter. The usage section will look like the following:

* We see the billing period to be `MONTHLY`
* We can see the reference to the `billingMeter#Code`
* We see one tier configured with a package size of 1 million.
* We see the price matrix, which in our case is set to be $2

[source,bash]
-----
"usages": [
  {
    "usageName": "usage-monthly-1",
    "usageType": "CONSUMABLE",
    "billingPeriod": "MONTHLY",
    "tiers": [
      {
        "tierNumber": 1,
        "blocks": [
          {
            "billingMeterCode": "input-M-tokens",
            "size": 1000000,
            "max": -1,
            "prices": [{"currency": "USD", "value": "2.0"}]
          }
        ]
      }
    ]
  }
]
-----

For reference, this uses the https://docs.killbill.io/latest/aviate-catalog-plugin[aviate catalog capabilities].

> The full paylaod and API call can be seens below:

Reviewer 2: ```suggestion
The full payload and API call can be seen below:
```

> Use aviate API: https://killbill.github.io/slate/aviate-catalog.html#create-plan-product-pricelist

Reviewer: ```suggestion
Use aviate API: https://apidocs.killbill.io/aviate-catalog#create-plan-product-pricelist
```
